builtin/providers/aws/resource_aws_waf_web_acl.go

```diff
@@ -0,0 +1,197 @@
+			"metric_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"rules": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action": &schema.Schema{
+							Type:     schema.TypeSet,
+							Required: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"type": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+						"priority": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"rule_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsWafWebAclCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	res, _ := conn.GetChangeToken(ct)
+
+	params := &waf.CreateWebACLInput{
+		ChangeToken:   res.ChangeToken,
+		DefaultAction: expandDefaultAction(d),
+		MetricName:    aws.String(d.Get("metric_name").(string)),
+		Name:          aws.String(d.Get("name").(string)),
+	}
+
+	resp, err := conn.CreateWebACL(params)
+	if err != nil {
+		return err
+	}
+	d.SetId(*resp.WebACL.WebACLId)
+	return resourceAwsWafWebAclUpdate(d, meta)
+}
+
+func resourceAwsWafWebAclRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+	params := &waf.GetWebACLInput{
+		WebACLId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetWebACL(params)
+	if err != nil {
+		return err
+	}
+
+	default_action := flattenDefaultAction(resp.WebACL.DefaultAction)
+	if default_action != nil {
+		if err := d.Set("default_action", default_action); err != nil {
+			return fmt.Errorf("error setting default_action: %s", err)
+		}
+	}
+	d.Set("name", resp.WebACL.Name)
+	d.Set("metric_name", resp.WebACL.MetricName)
+
+	return nil
+}
+
+func resourceAwsWafWebAclUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	resp, err := conn.GetChangeToken(ct)
+
+	req := &waf.UpdateWebACLInput{
+		ChangeToken: resp.ChangeToken,
+		WebACLId:    aws.String(d.Id()),
+	}
+
+	if d.HasChange("default_action") {
+		req.DefaultAction = expandDefaultAction(d)
+	}
+
+	_, err = conn.UpdateWebACL(req)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsWafWebAclRead(d, meta)
+}
+
+func resourceAwsWafWebAclDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafconn
+	// ChangeToken
+	var ct *waf.GetChangeTokenInput
+	resp, err := conn.GetChangeToken(ct)
+
+	req := &waf.DeleteWebACLInput{
+		ChangeToken: resp.ChangeToken,
+		WebACLId:    aws.String(d.Id()),
+	}
+	log.Printf("[INFO] Deleting WAF ACL")
+	_, err = conn.DeleteWebACL(req)
+
+	if err != nil {
```

Comment:
We should likely test here for whether the return code is a 404-equivalent and remove from state rather than erroring out if that is the case - this allows for us to track ACLs removed via the console.